Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1423 @@
+
+    init(previousState: ClientClosedServerOpenState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+      previousState.decompressor?.end()
+    }
+
+    init(previousState: ClientClosedServerIdleState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+      previousState.decompressor?.end()
+    }
+
+    init(previousState: ClientOpenServerClosedState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+      previousState.decompressor?.end()
+    }
+  }
+}
+
+// - MARK: Client
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct GRPCStreamStateMachine {
+  private var state: GRPCStreamStateMachineState
+  private var configuration: GRPCStreamStateMachineConfiguration
+  private var skipAssertions: Bool
+
+  init(
+    configuration: GRPCStreamStateMachineConfiguration,
+    maximumPayloadSize: Int,
+    skipAssertions: Bool = false
+  ) {
+    self.state = .clientIdleServerIdle(.init(maximumPayloadSize: maximumPayloadSize))
+    self.configuration = configuration
+    self.skipAssertions = skipAssertions
+  }
+
+  mutating func send(metadata: Metadata) throws -> HPACKHeaders {
+    switch self.configuration {
+    case .client(let clientConfiguration):
+      return try self.clientSend(metadata: metadata, configuration: clientConfiguration)
+    case .server(let serverConfiguration):
+      return try self.serverSend(metadata: metadata, configuration: serverConfiguration)
+    }
+  }
+
+  mutating func send(message: [UInt8], endStream: Bool) throws {
+    switch self.configuration {
+    case .client:
+      try self.clientSend(message: message, endStream: endStream)
+    case .server:
+      guard !endStream else {
+        throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+          "Can't end response stream by sending a message - send(status:metadata:trailersOnly:) must be called"
+        )
+      }
+      try self.serverSend(message: message)
+    }
+  }
+
+  mutating func send(status: Status, metadata: Metadata, trailersOnly: Bool) throws -> HPACKHeaders
+  {
+    switch self.configuration {
+    case .client:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client cannot send status and trailer."
+      )
+    case .server:
+      return try self.serverSend(
+        status: status,
+        customMetadata: metadata,
+        trailersOnly: trailersOnly
+      )
+    }
+  }
+
+  enum OnMetadataReceived: Equatable {
+    case receivedMetadata(Metadata)
+
+    // Client-specific actions
+    case failedRequest(Status)
+    case doNothing
+
+    // Server-specific actions
+    case rejectRPC(status: Status?, trailers: HPACKHeaders?)
```

Comment:
I was thinking the handler will call `send(status:metadata:)` afterwards when receiving this return value, so we need the status to be separate. It will be added as a header anyways in the call to `send(status:metadata:)`.

Reply:
I think it'd be simpler to skip the extra call into the state machine and just return the well formed `HPACKHeaders` from here instead.